### PR TITLE
Changeling refactor / doc / code improvements (kicks changeling out of `Life()`)

### PR DESCRIPTION
**Pull request by MrMelbert**
## About The Pull Request

- Auto-doc'd much of changeling code.
- Killed `handle_changeling` in `life()`. It's handled by signal now.
- Killed changeling code in `login()`. It's handled by signal now.
- Modernized a ton of code. Better var names, style guide compliance, etc.
- Split innate changeling abilities and purchased changeling abilities into two lists.
  - Changelings now don't have to delete and regain all of their basic abilities when respeccing. 
- Made modifying changeling chemicals uses a setter, `adjust_chemicals()`. 

## Why It's Good For The Game

Changeling code is very old and very stinky. This will hopefully make it not only easier to work with, but make behavior a bit more consistent and smooth.

## Changelog

:cl: Melbert
fix: You can no longer override the ling DNA you're currently disguised as. This was apparently always intended but broken for ages.
code: Modernized much of ling code.
refactor: Also refactored part of ling code. The ling chemical hud element should be a bit snappier now.
/:cl:

### CODE REVIEW

**Comment by JohnFulpWillard** [comment]
I tried talking about this a while ago
image

This was before ``COMSIG_LIVING_LIFE`` was even added, which itself was added by Floyd for basic mobs.
Kinda odd, just thought I'd mention.

**Comment by MrMelbert** [comment]
Interesting food for thought - i'd just want to be careful to not accidentally affect changeling balance by processing too quickly or too slowly in changing it to process instead of life.

File: code/modules/antagonists/changeling/cellular_emporium.dm [hunk 1]
```diff
@@ -25,35 +26,36 @@
...
-		var/req_dna = initial(ability.req_dna)
-		var/req_absorbs = initial(ability.req_absorbs)
-		AL["dna_cost"] = dna_cost
-		AL["can_purchase"] = ((req_absorbs <= true_absorbs) && (req_dna <= absorbed_dna_count) && (dna_cost <= genetic_points_remaining))
+		var/list/ability_data = list()
+		ability_data["name"] = initial(ability_path.name)
+		ability_data["desc"] = initial(ability_path.desc)
+		ability_data["path"] = ability_path
+		ability_data["helptext"] = initial(ability_path.helptext)
+		ability_data["owned"] = is_path_in_list_of_types(ability_path, changeling.purchased_powers)
```
Comment (LemonInTheDark) [inline code/modules/antagonists/changeling/cellular_emporium.dm:54]: This smells like a really kinda hacky way to go about things. Gut says we should just have this be a part of some global list or something.

Also why do you need to do istypes? this is basically is (type in list(objects)) right? couldn't you just maintain a list of purchased typepaths?

File: code/modules/antagonists/changeling/cellular_emporium.dm [hunk 2]
```diff
@@ -25,35 +26,36 @@
...
+		ability_data["dna_cost"] = dna_cost
+		ability_data["can_purchase"] = ((req_absorbs <= true_absorbs) && (req_dna <= absorbed_dna_count) && (dna_cost <= genetic_points_remaining))
```
Comment (LemonInTheDark) [inline code/modules/antagonists/changeling/cellular_emporium.dm:56]: ohhh brother. That's a lot to put on one line.
Maybe just like, suck it up and make it an if check?

File: code/__HELPERS/_lists.dm [hunk 3]
```diff
@@ -164,6 +166,17 @@
...
+ * Checks if any datum within [list_of_types] is of type [path_to_check].
+ */
+/proc/is_path_in_list_of_types(path_to_check, list/list_of_types)
+	if(!LAZYLEN(list_of_types) || !ispath(path_to_check))
+		return FALSE
+	for(var/thing in list_of_types)
+		if(istype(thing, path_to_check))
+			return TRUE
+	return FALSE
+
```
Comment (LemonInTheDark) [inline code/__HELPERS/_lists.dm:179]: This seems a bit off to me you know

File: code/modules/antagonists/changeling/changeling.dm [hunk 4]
```diff
@@ -74,324 +102,442 @@
...
+	RegisterSignal(living_mob, COMSIG_MOB_LOGIN, .proc/on_login)
+	RegisterSignal(living_mob, COMSIG_LIVING_LIFE, .proc/on_life)
+	living_mob.hud_used?.lingchemdisplay.invisibility = 0
+	living_mob.hud_used?.lingchemdisplay.maptext = FORMAT_CHEM_CHARGES_TEXT(chem_charges)
+
+	if(!iscarbon(mob_to_tweak))
+		return
+
+	var/mob/living/carbon/carbon_mob = mob_to_tweak
+	RegisterSignal(carbon_mob, list(COMSIG_MOB_MIDDLECLICKON, COMSIG_MOB_ALTCLICKON), .proc/on_click_sting)
```
Comment (LemonInTheDark) [inline code/modules/antagonists/changeling/changeling.dm:135]: are antagonist datums qdeleted? or are they like gasmixtures where we just free when we're done using them

File: code/modules/antagonists/changeling/changeling.dm [hunk 5]
```diff
@@ -74,324 +102,442 @@
...
+/datum/antagonist/changeling/proc/remove_changeling_powers(include_innate = FALSE)
+	if(!isliving(owner.current))
+		return
+
+	if(chosen_sting)
+		chosen_sting.unset_sting(owner.current)
+
+	for(var/datum/action/changeling/power as anything in purchased_powers)
+		purchased_powers -= power
+		power.Remove(owner.current)
```
Comment (LemonInTheDark) [inline code/modules/antagonists/changeling/changeling.dm:257]: This I think should just happen in action qdel. I think you could replace this whole forloop with just a QDEL_LIST macro

File: code/modules/antagonists/changeling/changeling.dm [hunk 6]
```diff
@@ -74,324 +102,442 @@
...
+		qdel(power)
+	if(include_innate)
+		for(var/datum/action/changeling/power as anything in innate_powers)
```
Comment (LemonInTheDark) [inline code/modules/antagonists/changeling/changeling.dm:260]: ditto here

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/modules/antagonists/changeling/changeling.dm (hunk 4 above)
Comment (MrMelbert) [reply code/modules/antagonists/changeling/changeling.dm:135]: antag datums self-qdel at the end of `/datum/antagonist/proc/on_removal()`

File: code/modules/antagonists/changeling/cellular_emporium.dm (hunk 1 above)
Comment (MrMelbert) [reply code/modules/antagonists/changeling/cellular_emporium.dm:54]: I could just make it an assoc list of path = instance, yeah

**Pull request merged**